packages/web-pkg/src/components/AppTopBar.vue

```diff
@@ -1,45 +1,130 @@
 <template>
-  <div class="oc-flex oc-p-s app-top-bar">
-    <oc-resource
-      v-if="resource"
-      id="app-top-bar-resource"
-      :is-thumbnail-displayed="false"
-      :resource="resource"
-    />
-    <div v-else />
-    <div>
-      <slot name="right" />
-      <oc-button
-        id="app-top-bar-close"
-        :aria-label="$gettext('Close')"
-        size="small"
-        @click="$emit('close')"
-      >
-        <oc-icon name="close" size="small" />
-      </oc-button>
+  <portal to="app.runtime.header.left">
+    <div class="oc-app-top-bar oc-flex">
+      <span class="oc-app-top-bar-inner oc-px-m oc-flex oc-flex-middle oc-flex-between">
+        <div class="open-file-bar oc-flex">
+          <oc-resource
+            v-if="resource"
+            id="app-top-bar-resource"
+            :is-thumbnail-displayed="false"
+            :resource="resource"
+          />
+        </div>
+
+        <div class="oc-flex main-actions">
+          <template v-if="mainActions.length && resource">
+            <context-action-menu
+              :menu-sections="[
+                {
+                  name: 'main-actions',
+                  items: mainActions
+                }
+              ]"
+              :action-options="{
+                resources: [resource]
+              }"
+              appearance="raw-inverse"
+              variation="brand"
+            />
+          </template>
+          <oc-button
+            id="app-top-bar-close"
+            v-oc-tooltip="closeButtonLabel"
+            appearance="raw-inverse"
+            variation="brand"
+            :aria-label="closeButtonLabel"
+            @click="$emit('close')"
+          >
+            <oc-icon name="close" size="small" />
+          </oc-button>
+        </div>
+      </span>
     </div>
-  </div>
+  </portal>
 </template>
 
 <script lang="ts">
 import { defineComponent, PropType } from 'vue'
 import { Resource } from 'web-client/src'
+import { Action } from '../composables/actions/types'
+import ContextActionMenu from 'web-pkg/src/components/ContextActions/ContextActionMenu.vue'
 
 export default defineComponent({
   name: 'AppTopBar',
+  components: {
+    ContextActionMenu
+  },
   props: {
+    mainActions: {
+      type: Array as PropType<Action[]>,
+      default: () => []
+    },
     resource: {
       type: Object as PropType<Resource>,
       default: null
     }
   },
-  emits: ['close']
+  emits: ['close'],
+  computed: {
+    contextMenuLabel() {
+      return this.$gettext('Show context menu')
+    },
+    closeButtonLabel() {
+      return this.$gettext('Close')
+    }
+  }
```

Comment:
Can we add new code in compostion api please?

Reply:
Done ;) thx